### ISSUE TITLE
putFile fails for Unicode file names (invalid OAuth signature)

### ISSUE THREAD

**Issue by thiloplanz** [evnm/dropbox-node#7]
I have a program working that does `putFile` successfully for most files, but files with Unicode file names fail with an OAuth signature validation error.

I have just updated to the latest version of everything (Node 0.4, OAuth 0.9.0, Dropbox 0.2.0, installed via npm), not sure if it worked before (with Node 0.2) or not.

If it matters, a file name that shows this error is データ同期プロトタイプ.pdf.
The expected signature base string has it as POST&https%3A%2F%2Fapi-content.dropbox.com%2F0%2Ffiles%2Fdropbox%2FDropSuite%2F&file%3Ddownloads%252F%25E3%2583%2587%25E3%2583%25BC%25E3%2582%25BF%25E5%2590%258C%25E6%259C%259F%25E3%2583%2597%25E3%2583%25AD%25E3%2583%2588%25E3%2582%25BF%25E3%2582%25A4%25E3%2583%2597.pdf%26oauth_consumer_key.......

Please let me know if there are any environment/locale settings that I might need to look out for (the file name prints fine in a UTF-8 console.info , though)

I'll check if the other API methods work for that file, and report back.

**Comment by evnm** [comment]
The issue seems to be with regard to the escaping of percent signs in the signature base string. Notice that the encoded URI in the error from Dropbox has a bunch of extraneous 25's in it (% in UTF-8). Dropbox seems to be expecting the file parameter in POST requests to be escaped twice.

I'm not yet sure if the issue is in dropbox-node, node-oauth, or the Dropbox API itself. Note that this bug should only be present in putFile, as the rest of the API methods are serviced via GET requests.

**Comment by thiloplanz** [comment]
> Note that this bug should only be present in putFile, as the rest of the API methods are serviced via GET requests.

I am afraid that is not the case, I get the same error for `getMetaData` as well.

Note that not all kinds of necessary encoding fails. File names with spaces in them for example work properly, even though they also contain % escapes.

**Comment by evnm** [comment]
Can you explain how you're getting it to fail on `getMetadata`? It works for me with dummy files with your example filename (データ同期プロトタイプ.pdf), %.txt, and others.

**Comment by thiloplanz** [comment]
> Can you explain how you're getting it to fail on `getMetadata`?

Sorry, that was my mistake. Part of my program runs in the browser (oauth.js by Netflix), part of it runs on the server (node.js). The errors for `getMetadata` happen with the in-browser part, the part that uses dropbox-node works fine (except for `putFile`).

**Comment by evnm** [comment]
This issue also appears to be present in other methods, including `createFolder` and `deleteItem`.

**Comment by linus** [comment]
I had issues even with spaces in filenames - that was fixed by escapePath(file):

```
self.oauth.post(CONTENT_API_URI + '/files/dropbox/' + path +
                '?file=' + escapePath(file)
                ...);
```

**Comment by evnm** [comment]
This issue appears to be fixed as of 0d609756c5.
